theano/gpuarray/c_code/dnn_sptf.c

```diff
@@ -0,0 +1,203 @@
+            "GpuDnnTransformerGradI: failed to allocate cuDNN tensor descriptor xdesc: %s",
+            cudnnGetErrorString( err ) );
+        FAIL;
+    }
+
+    err = cudnnCreateTensorDescriptor( &APPLY_SPECIFIC(ydesc) );
+    if ( err != CUDNN_STATUS_SUCCESS )
+    {
+        PyErr_Format( PyExc_MemoryError,
+            "GpuDnnTransformerGradI: failed to allocate cuDNN tensor descriptor ydesc: %s",
+            cudnnGetErrorString( err ) );
+        FAIL;
+    }
+}
+
+#section cleanup_code_struct
+
+if ( APPLY_SPECIFIC(xdesc) != NULL )
+    cudnnDestroyTensorDescriptor( APPLY_SPECIFIC(xdesc) );
+
+if ( APPLY_SPECIFIC(ydesc) != NULL )
+    cudnnDestroyTensorDescriptor( APPLY_SPECIFIC(ydesc) );
+
+#section support_code_struct
+
+int
+APPLY_SPECIFIC(dnn_sptf)(PyGpuArrayObject * input,
+                         PyGpuArrayObject * theta,
+                         PyArrayObject * grid_dims,
+                         cudnnSpatialTransformerDescriptor_t desc,
+                         PyGpuArrayObject ** output,
+                         PyGpuArrayObject ** grid,
+                         cudnnHandle_t _handle)
+{
+    PyGpuContextObject * gpu_ctx = input->context;
+    void * alpha_p;
+    void * beta_p;
+    double alpha = 1.0, beta = 0.0;
+    float af = alpha, bf = beta;
+    cudnnStatus_t err = CUDNN_STATUS_SUCCESS;
+    int num_images, num_channels, height, width;
+    size_t gpu_grid_dims[4], out_dims[4];
+
+    switch (input->ga.typecode)
+    {
+    case GA_DOUBLE:
+        alpha_p = (void *)&alpha;
+        beta_p = (void *)&beta;
+        break;
+    case GA_FLOAT:
+        alpha_p = (void *)&af;
+        beta_p = (void *)&bf;
+        break;
+    case GA_HALF:
+        alpha_p = (void *)&af;
+        beta_p = (void *)&bf;
+        break;
+    default:
+        PyErr_SetString( PyExc_TypeError,
+            "GpuDnnTransformer: unsupported type for input in spatial transformer." );
+        return 1;
+    }
+
+    if ( theta->ga.typecode != GA_FLOAT &&
+         theta->ga.typecode != GA_DOUBLE &&
+         theta->ga.typecode != GA_HALF )
+    {
+        PyErr_SetString( PyExc_TypeError,
+            "GpuDnnTransformer: unsupported data type for theta in spatial transformer." );
+        return 1;
+    }
+    else if ( PyGpuArray_DIM( theta, 1 ) != 2 && PyGpuArray_DIM( theta, 2 ) != 3 )
+    {
+        PyErr_Format( PyExc_RuntimeError,
+            "GpuDnnTransformer: incorrect dimensions for theta, expected (%d, %d, %d), got (%d, %d, %d)",
+            PyGpuArray_DIMS( theta )[0], 2, 3, PyGpuArray_DIMS( theta )[0],
+            PyGpuArray_DIMS( theta )[1], PyGpuArray_DIMS( theta )[2] );
+        return 1;
+    }
+
+    if ( PyArray_NDIM( grid_dims ) != 1 || PyArray_SIZE( grid_dims ) != 4 )
+    {
+        PyErr_SetString( PyExc_MemoryError,
+            "GpuDnnTransformer: grid_dims must have 4 elements." );
+        return 1;
+    }
+
+    // Obtain grid dimensions
+    num_images = (int) *( (npy_int *) PyArray_GETPTR1( grid_dims, 0 ) );
```

Comment:
As you have casted grid_dims to int64 in Python code, it should be `npy_int64` here and in next lines, not `npy_int`.